ndk/src/events/gift-wrapping.ts

```diff
@@ -0,0 +1,116 @@
+import { NDKEvent, NostrEvent } from "./index.js";
+import { NDKPrivateKeySigner } from "../signers/private-key";
+import { VerifiedEvent, getEventHash, verifiedSymbol } from "nostr-tools";
+import { EncryptionNip } from "./encryption.js";
+import { NDKUser } from "../user/index.js";
+import { NDKSigner } from "../signers/index.js";
+import { NDKKind } from "./kinds/index.js";
+
+// NIP 59 - adapted from Coracle
+
+export type GiftWrapParams = {
+    encryptionNip?: EncryptionNip
+    rumorKind?: number;
+    wrapTags?: string[][]
+}
+
+/**
+ * Instantiate a new (Nip59 gift wrapped) NDKEvent from any NDKevent
+ * @param this 
+ * @param recipient 
+ * @param signer 
+ * @param params 
+ * @returns 
+ */
+export async function giftWrap(this: NDKEvent, recipient: NDKUser, signer?: NDKSigner, params: GiftWrapParams = {}): Promise<NDKEvent>{
+  params.encryptionNip = params.encryptionNip || 'nip44';
+  if(!signer){
+      if(!this.ndk) 
+          throw new Error('no signer available for giftWrap')
+      signer = this.ndk.signer;
+  }
+  if(!signer) 
+      throw new Error('no signer')
+  if(!signer.encryptionEnabled || !signer.encryptionEnabled(params.encryptionNip)) 
+      throw new Error('signer is not able to giftWrap')
+  const rumor = getRumorEvent(this, params?.rumorKind)
+  const seal = await getSealEvent(rumor, recipient, signer, params.encryptionNip);
+  const wrap = await getWrapEvent(seal, recipient, params);
+  return new NDKEvent(this.ndk, wrap);
+}
+
+/**
+ * Instantiate a new (Nip59 un-wrapped rumor) NDKEvent from any gift wrapped NDKEvent
+ * @param this 
+ */
+export async function giftUnwrap(this:NDKEvent, sender?:NDKUser, signer?:NDKSigner, nip:EncryptionNip = 'nip44'): Promise<NDKEvent> {
+    sender = sender || new NDKUser({pubkey:this.pubkey})
+    if(!signer){
+        if(!this.ndk) 
+            throw new Error('no signer available for giftUnwrap')
+        signer = this.ndk.signer;
+    }
+    if(!signer) 
+        throw new Error('no signer')
+    try {
+      const seal = JSON.parse(await signer.decrypt(sender, this.content, nip)) as NostrEvent;
+      if (!seal) throw new Error("Failed to decrypt wrapper")
+
+      const rumorSender = new NDKUser({pubkey: seal.pubkey});
+      const rumor = JSON.parse(await signer.decrypt(rumorSender, seal.content, nip))
+      if (!rumor) throw new Error("Failed to decrypt seal")
+
+      if (seal.pubkey === rumor.pubkey) {
+        return new NDKEvent(this.ndk, rumor as NostrEvent)
```

Comment:
Important: need to manually perform a signature validation of the rumor otherwise anyone could fake these

Reply:
I guess you mean the seal event, right? The rumor must not be signed according to the nip. That's really important, good point.